### PR TITLE
[mlir][sparse] remove LevelType enum, construct LevelType from LevelF…

### PR DESCRIPTION
**Pull request by PeimingLiu**
…ormat and properties instead.

### CODE REVIEW

File: mlir/include/mlir/Dialect/SparseTensor/IR/Enums.h [hunk 1]
```diff
@@ -153,374 +153,246 @@ enum class Action : uint32_t {
...
+public:
+  /// Check that the `LevelType` contains a valid (possibly undefined) value.
+  static constexpr bool isValidLvlBits(uint64_t lvlBits) {
+    const uint64_t formatBits = lvlBits & 0xffff0000;
+    const uint64_t propertyBits = lvlBits & 0xffff;
+    // If undefined/dense/NOutOfM, then must be unique and ordered.
+    // Otherwise, the format must be one of the known ones.
+    return (formatBits <= 0x10000 || formatBits == 0x100000)
+               ? (propertyBits == 0)
+               : (formatBits == 0x20000 || formatBits == 0x40000 ||
```
Comment (aartbik) [inline mlir/include/mlir/Dialect/SparseTensor/IR/Enums.h:227]: This was there before, but shouldn't we have names for these formats?

File: mlir/include/mlir/Dialect/SparseTensor/IR/Enums.h [hunk 2]
```diff
@@ -153,374 +153,246 @@ enum class Action : uint32_t {
...
-/// Convert M of NOutOfM level type to the stored bits.
-constexpr uint64_t mToBits(uint64_t m) { return m << 40; }
+  /// Convert a LevelFormat to its corresponding LevelType with the given
+  /// properties. Returns std::nullopt when the properties are not applicable
+  /// for the input level format.
+  static std::optional<LevelType>
+  buildLvlType(LevelFormat lf,
+               const std::vector<LevelPropNonDefault> &properties,
+               uint64_t n = 0, uint64_t m = 0) {
+    uint64_t newN = n << 32;
```
Comment (aartbik) [inline mlir/include/mlir/Dialect/SparseTensor/IR/Enums.h:238]: to make sure we don't corrupt other bits, this should be (n & 0xff) << 32
(and perhaps an assert that n &0xff == n), same for m

File: mlir/include/mlir/Dialect/SparseTensor/IR/Enums.h [hunk 3]
```diff
@@ -153,374 +153,246 @@ enum class Action : uint32_t {
...
-/// Check if the `LevelType` is the special undefined value.
-constexpr bool isUndefLT(LevelType lt) { return lt == LevelType::Undef; }
+  LevelType stripProperties() const { return LevelType(lvlBits & ~0xffff); }
 
-/// Check if the `LevelType` is dense (regardless of properties).
-constexpr bool isDenseLT(LevelType lt) {
-  return (static_cast<uint64_t>(lt) & ~0xffff) ==
-         static_cast<uint64_t>(LevelType::Dense);
-}
+  /// Get N/M of NOutOfM level type.
```
Comment (aartbik) [inline mlir/include/mlir/Dialect/SparseTensor/IR/Enums.h:278]: Just dup this comment

/// Get N of NOutOfM level type.

/// Get M of NOutOfM level type.

File: mlir/include/mlir/Dialect/SparseTensor/IR/Enums.h [hunk 4]
```diff
@@ -153,374 +153,246 @@ enum class Action : uint32_t {
...
+}
+inline uint64_t getN(LevelType lt) { return lt.getN(); }
+inline uint64_t getM(LevelType lt) { return lt.getM(); }
+inline bool isValidNOutOfMLT(LevelType lt, uint64_t n, uint64_t m) {
+  return isNOutOfMLT(lt) && lt.getN() == n && lt.getM() == m;
+}
+inline std::string toMLIRString(LevelType lt) { return lt.toMLIRString(); }
 
 //
 // Ensure the above methods work as intended.
```
Comment (aartbik) [inline mlir/include/mlir/Dialect/SparseTensor/IR/Enums.h:393]: Does this comment apply to anything remaining? Seems the section is now empty?

**Review by aartbik — APPROVED** [review]

**Comment by joker-eph** [comment]
I had to revert in https://github.com/llvm/llvm-project/pull/81923 ; the MLIR gcc 7 bot was broken.

**Comment by PeimingLiu** [comment]
> I had to revert in #81923 ; the MLIR gcc 7 bot was broken.

Thanks!

pull/81934 should fix the error. Since I can not reproduce it locally, I will keep monitoring the nv-gcc7 buildbot after submission

**Pull request merged**